In [15]:
import numpy as np
import pandas as pd
from keras.models import Sequential
from  keras.layers import Dense
from  keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import LabelEncoder
from sklearn.pipeline import Pipeline
from sklearn.model_selection import StratifiedKFold
from sklearn.preprocessing import StandardScaler
from google.colab import files

In [5]:
file = files.upload()
np.random.seed(42)

Saving sonar.csv to sonar.csv


In [8]:
dataframe = pd.read_csv("/content/sonar.csv",header=None)
dataframe.head()

0       1       2       3       4   ...      56      57      58      59  60
0  0.0200  0.0371  0.0428  0.0207  0.0954  ...  0.0180  0.0084  0.0090  0.0032   R
1  0.0453  0.0523  0.0843  0.0689  0.1183  ...  0.0140  0.0049  0.0052  0.0044   R
2  0.0262  0.0582  0.1099  0.1083  0.0974  ...  0.0316  0.0164  0.0095  0.0078   R
3  0.0100  0.0171  0.0623  0.0205  0.0205  ...  0.0050  0.0044  0.0040  0.0117   R
4  0.0762  0.0666  0.0481  0.0394  0.0590  ...  0.0072  0.0048  0.0107  0.0094   R

[5 rows x 61 columns]

In [10]:
dataset = dataframe.values
x = dataset[:,0:60].astype(float)
y = dataset[:,60]

# Converting String to Binary

In [34]:
encoder = LabelEncoder()
encoder.fit(y)
encoded_y = encoder.transform(y)

# Neural Architecture


In [29]:
def create_baseline():
  model = Sequential()
  model.add(Dense(60,input_dim=60,kernel_initializer='normal',activation='relu'))
  model.add(Dense(1,kernel_initializer='normal',activation='sigmoid'))
  model.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])
  return model

# Evaluate model with standardized dataset

In [31]:
estimator = KerasClassifier(build_fn=create_baseline, epochs=100, batch_size=5, verbose=0)
kfold = StratifiedKFold(n_splits=10, shuffle=True, random_state=42)
results = cross_val_score(estimator, x, encoded_y, cv=kfold)
print("Baseline: %.2f%% (%.2f%%)" %(results.mean()*100, results.std()*100))

<ipython-input-31-d43d1a0704cf>:1: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  estimator = KerasClassifier(build_fn=create_baseline, epochs=100, batch_size=5, verbose=0)


Baseline: 80.31% (11.17%)


Baseline: 80.31% (11.17%)
This is an excellent score without doing any hard work.

In [33]:
estimators = []
estimators.append(('Standardize',StandardScaler()))
estimators.append(('mlp',KerasClassifier(build_fn=create_baseline,epochs=100,batch_size=5,verbose=0))
)
pipeline = Pipeline(estimators)
kfold = StratifiedKFold(n_splits=10,shuffle=True,random_state=42)
results = cross_val_score(pipeline,x,encoded_y,cv=kfold)
print("Standardized: %.2f%% (%.2f%%)" % (results.mean()*100, results.std()*100))

<ipython-input-33-6f05e66bfc2a>:3: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  estimators.append(('mlp',KerasClassifier(build_fn=create_baseline,epochs=100,batch_size=5,verbose=0))


Standardized: 86.07% (6.86%)


# Feature Extraction - Reducing the neurons in the model architecture

In [35]:
def create_baseline():
  model = Sequential()
  model.add(Dense(30,input_dim=60,kernel_initializer='normal',activation='relu'))
  model.add(Dense(1,kernel_initializer='normal',activation='sigmoid'))
  model.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])
  return model
estimators = []
estimators.append(('Standardize',StandardScaler()))
estimators.append(('mlp',KerasClassifier(build_fn=create_baseline,epochs=100,batch_size=5,verbose=0))
)
pipeline = Pipeline(estimators)
kfold = StratifiedKFold(n_splits=10,shuffle=True,random_state=42)
results = cross_val_score(pipeline,x,encoded_y,cv=kfold)
print("Standardized: %.2f%% (%.2f%%)" % (results.mean()*100, results.std()*100))

<ipython-input-35-67e2110c461c>:9: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  estimators.append(('mlp',KerasClassifier(build_fn=create_baseline,epochs=100,batch_size=5,verbose=0))


Standardized: 86.52% (7.67%)


This is a great result because we are
doing slightly better with a network half the size, which in turn takes half the time to train.

# Use a Larger Network Topology

In [38]:
def create_baseline():
  model = Sequential()
  model.add(Dense(60,input_dim=60,kernel_initializer='normal',activation='relu'))
  model.add(Dense(30,kernel_initializer='normal',activation='relu'))
  model.add(Dense(1,kernel_initializer='normal',activation='sigmoid'))
  model.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])
  return model
estimators = []
estimators.append(('Standardize',StandardScaler()))
estimators.append(('mlp',KerasClassifier(build_fn=create_baseline,epochs=100,batch_size=5,verbose=0))
)
pipeline = Pipeline(estimators)
kfold = StratifiedKFold(n_splits=10,shuffle=True,random_state=42)
results = cross_val_score(pipeline,x,encoded_y,cv=kfold)
print("Standardized: %.2f%% (%.2f%%)" % (results.mean()*100, results.std()*100))

<ipython-input-38-b13ff493a526>:10: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  estimators.append(('mlp',KerasClassifier(build_fn=create_baseline,epochs=100,batch_size=5,verbose=0))


Standardized: 86.05% (9.90%)
